Project sync takes significantly longer after upgrade to Gradle 5.0 when certain libraries are used

After upgrading Gradle from 4.10.2 to 5.0 project sync takes significantly longer when certain libraries are used from Google Maven Repository.

Build > Sync window shows lots of errors like these:
<img width="1225" alt="screenshot 2018-11-28 at 20 27 57" src="https://user-images.githubusercontent.com/119192/49169989-26463700-f34c-11e8-9a1c-4dc96f1ea52e.png">

<details>
<summary>Stacktrace</summary>

```
org.gradle.api.resources.ResourceException: Could not get resource 'https://dl.google.com/dl/android/maven2/com/google/firebase/firebase-core/16.0.5/firebase-core-16.0.5-javadoc.jar'.
	at org.gradle.internal.resource.ResourceExceptions.failure(ResourceExceptions.java:74)
	at org.gradle.internal.resource.ResourceExceptions.getFailed(ResourceExceptions.java:57)
	at org.gradle.internal.resource.transfer.AccessorBackedExternalResource.withContentIfPresent(AccessorBackedExternalResource.java:146)
	at org.gradle.internal.resource.BuildOperationFiringExternalResourceDecorator$11.call(BuildOperationFiringExternalResourceDecorator.java:237)
	at org.gradle.internal.resource.BuildOperationFiringExternalResourceDecorator$11.call(BuildOperationFiringExternalResourceDecorator.java:229)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:315)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:305)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:175)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:101)
	at org.gradle.internal.operations.DelegatingBuildOperationExecutor.call(DelegatingBuildOperationExecutor.java:36)
	at org.gradle.internal.resource.BuildOperationFiringExternalResourceDecorator.withContentIfPresent(BuildOperationFiringExternalResourceDecorator.java:229)
	at org.gradle.internal.resource.transfer.DefaultCacheAwareExternalResourceAccessor.copyToCache(DefaultCacheAwareExternalResourceAccessor.java:199)
	at org.gradle.internal.resource.transfer.DefaultCacheAwareExternalResourceAccessor.access$300(DefaultCacheAwareExternalResourceAccessor.java:54)
	at org.gradle.internal.resource.transfer.DefaultCacheAwareExternalResourceAccessor$1.create(DefaultCacheAwareExternalResourceAccessor.java:89)
	at org.gradle.internal.resource.transfer.DefaultCacheAwareExternalResourceAccessor$1.create(DefaultCacheAwareExternalResourceAccessor.java:81)
	at org.gradle.cache.internal.ProducerGuard$AdaptiveProducerGuard.guardByKey(ProducerGuard.java:97)
	at org.gradle.internal.resource.transfer.DefaultCacheAwareExternalResourceAccessor.getResource(DefaultCacheAwareExternalResourceAccessor.java:81)
	at org.gradle.api.internal.artifacts.repositories.resolver.DefaultExternalResourceArtifactResolver.downloadByCoords(DefaultExternalResourceArtifactResolver.java:133)
	at org.gradle.api.internal.artifacts.repositories.resolver.DefaultExternalResourceArtifactResolver.downloadStaticResource(DefaultExternalResourceArtifactResolver.java:97)
	at org.gradle.api.internal.artifacts.repositories.resolver.DefaultExternalResourceArtifactResolver.resolveArtifact(DefaultExternalResourceArtifactResolver.java:66)
	at org.gradle.api.internal.artifacts.repositories.resolver.ExternalResourceResolver$RemoteRepositoryAccess.resolveArtifact(ExternalResourceResolver.java:477)
	at org.gradle.api.internal.artifacts.ivyservice.ivyresolve.CachingModuleComponentRepository$ResolveAndCacheRepositoryAccess.resolveArtifact(CachingModuleComponentRepository.java:424)
	at org.gradle.api.internal.artifacts.ivyservice.ivyresolve.ErrorHandlingModuleComponentRepository$ErrorHandlingModuleComponentRepositoryAccess.lambda$resolveArtifact$12(ErrorHandlingModuleComponentRepository.java:180)
	at org.gradle.api.internal.artifacts.ivyservice.ivyresolve.ErrorHandlingModuleComponentRepository$ErrorHandlingModuleComponentRepositoryAccess.tryResolveAndMaybeBlacklist(ErrorHandlingModuleComponentRepository.java:236)
	at org.gradle.api.internal.artifacts.ivyservice.ivyresolve.ErrorHandlingModuleComponentRepository$ErrorHandlingModuleComponentRepositoryAccess.performOperationWithRetries(ErrorHandlingModuleComponentRepository.java:202)
	at org.gradle.api.internal.artifacts.ivyservice.ivyresolve.ErrorHandlingModuleComponentRepository$ErrorHandlingModuleComponentRepositoryAccess.resolveArtifact(ErrorHandlingModuleComponentRepository.java:178)
	at org.gradle.api.internal.artifacts.ivyservice.ivyresolve.RepositoryChainArtifactResolver.resolveArtifact(RepositoryChainArtifactResolver.java:87)
	at org.gradle.api.internal.artifacts.ivyservice.ivyresolve.ErrorHandlingArtifactResolver.resolveArtifact(ErrorHandlingArtifactResolver.java:46)
	at org.gradle.api.internal.artifacts.query.DefaultArtifactResolutionQuery.addArtifacts(DefaultArtifactResolutionQuery.java:179)
	at org.gradle.api.internal.artifacts.query.DefaultArtifactResolutionQuery.buildComponentResult(DefaultArtifactResolutionQuery.java:168)
	at org.gradle.api.internal.artifacts.query.DefaultArtifactResolutionQuery.createResult(DefaultArtifactResolutionQuery.java:142)
	at org.gradle.api.internal.artifacts.query.DefaultArtifactResolutionQuery.execute(DefaultArtifactResolutionQuery.java:133)
	at com.android.build.gradle.internal.ide.dependencies.ArtifactDependencyGraph.handleSources(ArtifactDependencyGraph.java:351)
	at com.android.build.gradle.internal.ide.dependencies.ArtifactDependencyGraph.createDependencies(ArtifactDependencyGraph.java:309)
	at com.android.build.gradle.internal.ide.ModelBuilder.getDependencies(ModelBuilder.java:815)
	at com.android.build.gradle.internal.ide.ModelBuilder.createAndroidArtifact(ModelBuilder.java:919)
	at com.android.build.gradle.internal.ide.ModelBuilder.createVariant(ModelBuilder.java:583)
	at com.android.build.gradle.internal.ide.ModelBuilder.buildAndroidProject(ModelBuilder.java:399)
	at com.android.build.gradle.internal.ide.ModelBuilder.buildAll(ModelBuilder.java:212)
	at com.android.build.gradle.internal.AppModelBuilder.buildAll(AppModelBuilder.kt:71)
	at org.gradle.tooling.provider.model.internal.DefaultToolingModelBuilderRegistry$BuildOperationWrappingToolingModelBuilder$1$1.create(DefaultToolingModelBuilderRegistry.java:102)
	at org.gradle.api.internal.project.DefaultProjectStateRegistry.withLenientState(DefaultProjectStateRegistry.java:132)
	at org.gradle.tooling.provider.model.internal.DefaultToolingModelBuilderRegistry$BuildOperationWrappingToolingModelBuilder$1.call(DefaultToolingModelBuilderRegistry.java:98)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:315)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:305)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:175)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:101)
	at org.gradle.internal.operations.DelegatingBuildOperationExecutor.call(DelegatingBuildOperationExecutor.java:36)
	at org.gradle.tooling.provider.model.internal.DefaultToolingModelBuilderRegistry$BuildOperationWrappingToolingModelBuilder.buildAll(DefaultToolingModelBuilderRegistry.java:95)
	at org.gradle.tooling.internal.provider.runner.DefaultBuildController.getModel(DefaultBuildController.java:79)
	at org.gradle.tooling.internal.consumer.connection.InternalBuildActionAdapter$2.getModel(InternalBuildActionAdapter.java:77)
	at org.gradle.tooling.internal.consumer.connection.BuildControllerAdapter.getModel(BuildControllerAdapter.java:62)
	at org.gradle.tooling.internal.consumer.connection.AbstractBuildController.findModel(AbstractBuildController.java:57)
	at org.gradle.tooling.internal.consumer.connection.AbstractBuildController.findModel(AbstractBuildController.java:44)
	at org.jetbrains.plugins.gradle.model.ProjectImportAction.addExtraProject(ProjectImportAction.java:143)
	at org.jetbrains.plugins.gradle.model.ProjectImportAction.execute(ProjectImportAction.java:84)
	at org.jetbrains.plugins.gradle.model.ProjectImportAction.execute(ProjectImportAction.java:40)
	at org.gradle.tooling.internal.consumer.connection.InternalBuildActionAdapter.execute(InternalBuildActionAdapter.java:80)
	at org.gradle.tooling.internal.provider.runner.ClientProvidedBuildActionRunner.buildResult(ClientProvidedBuildActionRunner.java:81)
	at org.gradle.tooling.internal.provider.runner.ClientProvidedBuildActionRunner.access$100(ClientProvidedBuildActionRunner.java:38)
	at org.gradle.tooling.internal.provider.runner.ClientProvidedBuildActionRunner$1.buildFinished(ClientProvidedBuildActionRunner.java:62)
	at sun.reflect.GeneratedMethodAccessor137.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.event.DefaultListenerManager$ListenerDetails.dispatch(DefaultListenerManager.java:376)
	at org.gradle.internal.event.DefaultListenerManager$ListenerDetails.dispatch(DefaultListenerManager.java:358)
	at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:58)
	at org.gradle.internal.event.DefaultListenerManager$EventBroadcast$ListenerDispatch.dispatch(DefaultListenerManager.java:346)
	at org.gradle.internal.event.DefaultListenerManager$EventBroadcast$ListenerDispatch.dispatch(DefaultListenerManager.java:333)
	at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:42)
	at org.gradle.internal.event.BroadcastDispatch$SingletonDispatch.dispatch(BroadcastDispatch.java:230)
	at org.gradle.internal.event.BroadcastDispatch$SingletonDispatch.dispatch(BroadcastDispatch.java:149)
	at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:58)
	at org.gradle.internal.event.BroadcastDispatch$CompositeDispatch.dispatch(BroadcastDispatch.java:324)
	at org.gradle.internal.event.BroadcastDispatch$CompositeDispatch.dispatch(BroadcastDispatch.java:234)
	at org.gradle.internal.event.ListenerBroadcast.dispatch(ListenerBroadcast.java:140)
	at org.gradle.internal.event.ListenerBroadcast.dispatch(ListenerBroadcast.java:37)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:93)
	at com.sun.proxy.$Proxy13.buildFinished(Unknown Source)
	at org.gradle.initialization.DefaultGradleLauncher.finishBuild(DefaultGradleLauncher.java:177)
	at org.gradle.initialization.DefaultGradleLauncher.finishBuild(DefaultGradleLauncher.java:140)
	at org.gradle.internal.invocation.GradleBuildController$3.create(GradleBuildController.java:104)
	at org.gradle.internal.invocation.GradleBuildController$3.create(GradleBuildController.java:96)
	at org.gradle.internal.work.DefaultWorkerLeaseService.withLocks(DefaultWorkerLeaseService.java:183)
	at org.gradle.internal.work.StopShieldingWorkerLeaseService.withLocks(StopShieldingWorkerLeaseService.java:40)
	at org.gradle.internal.invocation.GradleBuildController.doBuild(GradleBuildController.java:96)
	at org.gradle.internal.invocation.GradleBuildController.configure(GradleBuildController.java:85)
	at org.gradle.tooling.internal.provider.runner.ClientProvidedBuildActionRunner.run(ClientProvidedBuildActionRunner.java:70)
	at org.gradle.launcher.exec.ChainingBuildActionRunner.run(ChainingBuildActionRunner.java:35)
	at org.gradle.launcher.exec.ChainingBuildActionRunner.run(ChainingBuildActionRunner.java:35)
	at org.gradle.tooling.internal.provider.ValidatingBuildActionRunner.run(ValidatingBuildActionRunner.java:32)
	at org.gradle.launcher.exec.RunAsBuildOperationBuildActionRunner$3.run(RunAsBuildOperationBuildActionRunner.java:49)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:301)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:293)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:175)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:91)
	at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
	at org.gradle.launcher.exec.RunAsBuildOperationBuildActionRunner.run(RunAsBuildOperationBuildActionRunner.java:44)
	at org.gradle.tooling.internal.provider.SubscribableBuildActionRunner.run(SubscribableBuildActionRunner.java:51)
	at org.gradle.launcher.exec.InProcessBuildActionExecuter$1.transform(InProcessBuildActionExecuter.java:47)
	at org.gradle.launcher.exec.InProcessBuildActionExecuter$1.transform(InProcessBuildActionExecuter.java:44)
	at org.gradle.composite.internal.DefaultRootBuildState.run(DefaultRootBuildState.java:79)
	at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:44)
	at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:30)
	at org.gradle.launcher.exec.BuildTreeScopeBuildActionExecuter.execute(BuildTreeScopeBuildActionExecuter.java:39)
	at org.gradle.launcher.exec.BuildTreeScopeBuildActionExecuter.execute(BuildTreeScopeBuildActionExecuter.java:25)
	at org.gradle.tooling.internal.provider.ContinuousBuildActionExecuter.execute(ContinuousBuildActionExecuter.java:78)
	at org.gradle.tooling.internal.provider.ContinuousBuildActionExecuter.execute(ContinuousBuildActionExecuter.java:52)
	at org.gradle.tooling.internal.provider.ServicesSetupBuildActionExecuter.execute(ServicesSetupBuildActionExecuter.java:62)
	at org.gradle.tooling.internal.provider.ServicesSetupBuildActionExecuter.execute(ServicesSetupBuildActionExecuter.java:34)
	at org.gradle.tooling.internal.provider.GradleThreadBuildActionExecuter.execute(GradleThreadBuildActionExecuter.java:36)
	at org.gradle.tooling.internal.provider.GradleThreadBuildActionExecuter.execute(GradleThreadBuildActionExecuter.java:25)
	at org.gradle.tooling.internal.provider.ParallelismConfigurationBuildActionExecuter.execute(ParallelismConfigurationBuildActionExecuter.java:43)
	at org.gradle.tooling.internal.provider.ParallelismConfigurationBuildActionExecuter.execute(ParallelismConfigurationBuildActionExecuter.java:29)
	at org.gradle.tooling.internal.provider.StartParamsValidatingActionExecuter.execute(StartParamsValidatingActionExecuter.java:59)
	at org.gradle.tooling.internal.provider.StartParamsValidatingActionExecuter.execute(StartParamsValidatingActionExecuter.java:31)
	at org.gradle.tooling.internal.provider.SessionFailureReportingActionExecuter.execute(SessionFailureReportingActionExecuter.java:59)
	at org.gradle.tooling.internal.provider.SessionFailureReportingActionExecuter.execute(SessionFailureReportingActionExecuter.java:44)
	at org.gradle.tooling.internal.provider.SetupLoggingActionExecuter.execute(SetupLoggingActionExecuter.java:46)
	at org.gradle.tooling.internal.provider.SetupLoggingActionExecuter.execute(SetupLoggingActionExecuter.java:30)
	at org.gradle.launcher.daemon.server.exec.ExecuteBuild.doBuild(ExecuteBuild.java:67)
	at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:36)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:122)
	at org.gradle.launcher.daemon.server.exec.WatchForDisconnection.execute(WatchForDisconnection.java:37)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:122)
	at org.gradle.launcher.daemon.server.exec.ResetDeprecationLogger.execute(ResetDeprecationLogger.java:26)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:122)
	at org.gradle.launcher.daemon.server.exec.RequestStopIfSingleUsedDaemon.execute(RequestStopIfSingleUsedDaemon.java:34)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:122)
	at org.gradle.launcher.daemon.server.exec.ForwardClientInput$2.call(ForwardClientInput.java:74)
	at org.gradle.launcher.daemon.server.exec.ForwardClientInput$2.call(ForwardClientInput.java:72)
	at org.gradle.util.Swapper.swap(Swapper.java:38)
	at org.gradle.launcher.daemon.server.exec.ForwardClientInput.execute(ForwardClientInput.java:72)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:122)
	at org.gradle.launcher.daemon.server.exec.LogAndCheckHealth.execute(LogAndCheckHealth.java:55)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:122)
	at org.gradle.launcher.daemon.server.exec.LogToClient.doBuild(LogToClient.java:62)
	at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:36)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:122)
	at org.gradle.launcher.daemon.server.exec.EstablishBuildEnvironment.doBuild(EstablishBuildEnvironment.java:81)
	at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:36)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:122)
	at org.gradle.launcher.daemon.server.exec.StartBuildOrRespondWithBusy$1.run(StartBuildOrRespondWithBusy.java:50)
	at org.gradle.launcher.daemon.server.DaemonStateCoordinator$1.run(DaemonStateCoordinator.java:295)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.io.IOException: Response 204: No Content has no content!
	at org.gradle.internal.resource.transport.http.HttpClientResponse.getContent(HttpClientResponse.java:51)
	at org.gradle.internal.resource.transport.http.HttpResponseResource.openStream(HttpResponseResource.java:109)
	at org.gradle.internal.resource.transfer.ProgressLoggingExternalResourceAccessor$ProgressLoggingExternalResource.openStream(ProgressLoggingExternalResourceAccessor.java:62)
	at org.gradle.internal.resource.transfer.AccessorBackedExternalResource.withContentIfPresent(AccessorBackedExternalResource.java:135)
	... 149 more
```
</details>
<br>
I'm using Android Studio 3.3 RC1 but I've also reproduced this in 3.4 Preview 5.

### Steps to reproduce
1. `git clone https://github.com/technoir42/gradle-5.0-test.git`
2. Open the cloned project in Android Studio.
3. Project sync takes 30-40 seconds.
4. Comment-out Firebase libraries in `app/build.gradle`.
5. Sync Project with Gradle.
6. Sync now takes only 1-2 seconds.

**To compare with Gradle 4.10.x:**
1. Checkout branch `gradle-4.10` of the same repository.
2. Open the project in Android Studio.
3. Project sync takes only 1-3 seconds.

### Environment
macOS Mojave 10.14.1
Oracle JDK 8u191

--- a/subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependencyUnresolvedModuleIntegrationTest.groovy
+++ b/subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependencyUnresolvedModuleIntegrationTest.groovy
@@ -84,9 +84,13 @@ class DependencyUnresolvedModuleIntegrationTest extends AbstractHttpDependencyRe
     }
 
     @Unroll
-    def "fails single application dependency resolution if #protocol connection exceeds timeout"() {
+    def "fails single application dependency resolution if #protocol connection exceeds timeout (retries = #maxRetries)"() {
+        maxHttpRetries = maxRetries
+
         given:
-        blockingForProtocol(protocol, moduleA.pom)
+        maxHttpRetries.times {
+            blockingForProtocol(protocol, moduleA.pom)
+        }
         buildFile << """
             ${mavenRepository(mavenHttpRepo)}
             ${customConfigDependencyAssignment(moduleA)}
@@ -101,7 +105,11 @@ class DependencyUnresolvedModuleIntegrationTest extends AbstractHttpDependencyRe
         !downloadedLibsDir.isDirectory()
 
         where:
-        protocol << ['http', 'https']
+        protocol | maxRetries
+        'http'   | 1
+        'https'  | 1
+        'http'   | 2
+        'https'  | 2
     }
 
     @Unroll

--- a/subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/http/HttpAuthenticationDependencyResolutionIntegrationTest.groovy
+++ b/subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/http/HttpAuthenticationDependencyResolutionIntegrationTest.groovy
@@ -32,6 +32,11 @@ import static org.gradle.test.fixtures.server.http.AuthScheme.NTLM
 class HttpAuthenticationDependencyResolutionIntegrationTest extends AbstractHttpDependencyResolutionTest {
     static String badCredentials = "credentials{username 'testuser'; password 'bad'}"
 
+    def setup() {
+        // by setting this to >1, we assert that an authentication error is NOT going to cause retries
+        maxHttpRetries = 3
+    }
+
     @Unroll
     def "can resolve dependencies using #authSchemeName scheme from #authScheme authenticated HTTP ivy repository"() {
         given:

--- a/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/ErrorHandlingModuleComponentRepository.java
+++ b/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/ErrorHandlingModuleComponentRepository.java
@@ -16,6 +16,8 @@
 
 package org.gradle.api.internal.artifacts.ivyservice.ivyresolve;
 
+import org.apache.http.HttpStatus;
+import org.apache.http.conn.HttpHostConnectException;
 import org.gradle.api.Transformer;
 import org.gradle.api.artifacts.ComponentMetadataSupplierDetails;
 import org.gradle.api.artifacts.component.ComponentArtifactIdentifier;
@@ -34,6 +36,7 @@ import org.gradle.internal.component.model.ComponentArtifactMetadata;
 import org.gradle.internal.component.model.ComponentOverrideMetadata;
 import org.gradle.internal.component.model.ComponentResolveMetadata;
 import org.gradle.internal.component.model.ModuleSource;
+import org.gradle.internal.exceptions.DefaultMultiCauseException;
 import org.gradle.internal.resolve.ArtifactNotFoundException;
 import org.gradle.internal.resolve.ArtifactResolveException;
 import org.gradle.internal.resolve.ModuleVersionResolveException;
@@ -43,7 +46,10 @@ import org.gradle.internal.resolve.result.BuildableComponentArtifactsResolveResu
 import org.gradle.internal.resolve.result.BuildableModuleComponentMetaDataResolveResult;
 import org.gradle.internal.resolve.result.BuildableModuleVersionListingResolveResult;
 import org.gradle.internal.resolve.result.ErroringResolveResult;
+import org.gradle.internal.resource.transport.http.HttpErrorStatusCodeException;
 
+import java.net.SocketTimeoutException;
+import java.util.List;
 import java.util.Map;
 import java.util.concurrent.Callable;
 
@@ -138,56 +144,56 @@ public class ErrorHandlingModuleComponentRepository implements ModuleComponentRe
         @Override
         public void listModuleVersions(ModuleDependencyMetadata dependency, BuildableModuleVersionListingResolveResult result) {
             performOperationWithRetries(result,
-                () -> delegate.listModuleVersions(dependency, result),
-                () -> new ModuleVersionResolveException(dependency.getSelector(), BLACKLISTED_REPOSITORY_ERROR_MESSAGE),
-                throwable -> {
-                    ModuleComponentSelector selector = dependency.getSelector();
-                    String message = "Failed to list versions for " + selector.getGroup() + ":" + selector.getModule() + ".";
-                    return new ModuleVersionResolveException(selector, message, throwable);
-                });
+                    () -> delegate.listModuleVersions(dependency, result),
+                    () -> new ModuleVersionResolveException(dependency.getSelector(), BLACKLISTED_REPOSITORY_ERROR_MESSAGE),
+                    throwable -> {
+                        ModuleComponentSelector selector = dependency.getSelector();
+                        String message = "Failed to list versions for " + selector.getGroup() + ":" + selector.getModule() + ".";
+                        return new ModuleVersionResolveException(selector, message, throwable);
+                    });
         }
 
         @Override
         public void resolveComponentMetaData(ModuleComponentIdentifier moduleComponentIdentifier, ComponentOverrideMetadata requestMetaData, BuildableModuleComponentMetaDataResolveResult result) {
             performOperationWithRetries(result,
-                () -> delegate.resolveComponentMetaData(moduleComponentIdentifier, requestMetaData, result),
-                () -> new ModuleVersionResolveException(moduleComponentIdentifier, BLACKLISTED_REPOSITORY_ERROR_MESSAGE),
-                throwable -> new ModuleVersionResolveException(moduleComponentIdentifier, throwable)
+                    () -> delegate.resolveComponentMetaData(moduleComponentIdentifier, requestMetaData, result),
+                    () -> new ModuleVersionResolveException(moduleComponentIdentifier, BLACKLISTED_REPOSITORY_ERROR_MESSAGE),
+                    throwable -> new ModuleVersionResolveException(moduleComponentIdentifier, throwable)
             );
         }
 
         @Override
         public void resolveArtifactsWithType(ComponentResolveMetadata component, ArtifactType artifactType, BuildableArtifactSetResolveResult result) {
             performOperationWithRetries(result,
-                () -> delegate.resolveArtifactsWithType(component, artifactType, result),
-                () -> new ArtifactResolveException(component.getId(), BLACKLISTED_REPOSITORY_ERROR_MESSAGE),
-                throwable -> new ArtifactResolveException(component.getId(), throwable)
+                    () -> delegate.resolveArtifactsWithType(component, artifactType, result),
+                    () -> new ArtifactResolveException(component.getId(), BLACKLISTED_REPOSITORY_ERROR_MESSAGE),
+                    throwable -> new ArtifactResolveException(component.getId(), throwable)
             );
         }
 
         @Override
         public void resolveArtifacts(ComponentResolveMetadata component, BuildableComponentArtifactsResolveResult result) {
             performOperationWithRetries(result,
-                () -> delegate.resolveArtifacts(component, result),
-                () -> new ArtifactResolveException(component.getId(), BLACKLISTED_REPOSITORY_ERROR_MESSAGE),
-                throwable -> new ArtifactResolveException(component.getId(), throwable));
+                    () -> delegate.resolveArtifacts(component, result),
+                    () -> new ArtifactResolveException(component.getId(), BLACKLISTED_REPOSITORY_ERROR_MESSAGE),
+                    throwable -> new ArtifactResolveException(component.getId(), throwable));
         }
 
         @Override
         public void resolveArtifact(ComponentArtifactMetadata artifact, ModuleSource moduleSource, BuildableArtifactResolveResult result) {
             performOperationWithRetries(result,
-                () -> {
-                    delegate.resolveArtifact(artifact, moduleSource, result);
-                    if (result.hasResult()) {
-                        ArtifactResolveException failure = result.getFailure();
-                        if (!(failure instanceof ArtifactNotFoundException)) {
-                            return failure;
+                    () -> {
+                        delegate.resolveArtifact(artifact, moduleSource, result);
+                        if (result.hasResult()) {
+                            ArtifactResolveException failure = result.getFailure();
+                            if (!(failure instanceof ArtifactNotFoundException)) {
+                                return failure;
+                            }
                         }
-                    }
-                    return null;
-                },
-                () -> new ArtifactResolveException(artifact.getId(), BLACKLISTED_REPOSITORY_ERROR_MESSAGE),
-                throwable -> new ArtifactResolveException(artifact.getId(), throwable));
+                        return null;
+                    },
+                    () -> new ArtifactResolveException(artifact.getId(), BLACKLISTED_REPOSITORY_ERROR_MESSAGE),
+                    throwable -> new ArtifactResolveException(artifact.getId(), throwable));
         }
 
         private <E extends Throwable, R extends ErroringResolveResult<E>> void performOperationWithRetries(R result,
@@ -244,7 +250,8 @@ public class ErrorHandlingModuleComponentRepository implements ModuleComponentRe
                     unexpectedFailure = throwable;
                     failure = onError.transform(throwable);
                 }
-                if (retries == maxTentativesCount) {
+                boolean doNotRetry = !isLikelyTransientNetworkingIssue(failure);
+                if (doNotRetry || retries == maxTentativesCount) {
                     if (unexpectedFailure != null) {
                         repositoryBlacklister.blacklistRepository(repositoryId, unexpectedFailure);
                     }
@@ -260,6 +267,42 @@ public class ErrorHandlingModuleComponentRepository implements ModuleComponentRe
                     }
                 }
             }
+        }
+
+        /**
+         * Determines if an error should cause a retry. We will currently retry:
+         * <ul>
+         * <li>on a network timeout</li>
+         * <li>on a server error (return code 5xx)</li>
+         * <li>on rate limiting</li>
+         * </ul>
+         */
+        private static <E extends Throwable> boolean isLikelyTransientNetworkingIssue(E failure) {
+            if (failure instanceof SocketTimeoutException || failure instanceof HttpHostConnectException) {
+                return true;
+            }
+            if (failure instanceof DefaultMultiCauseException) {
+                List<? extends Throwable> causes = ((DefaultMultiCauseException) failure).getCauses();
+                for (Throwable cause : causes) {
+                    if (isLikelyTransientNetworkingIssue(cause)) {
+                        return true;
+                    }
+                }
+            }
+            if (failure instanceof HttpErrorStatusCodeException) {
+                HttpErrorStatusCodeException httpError = (HttpErrorStatusCodeException) failure;
+                return httpError.isServerError() || isTransientClientError(httpError.getStatusCode());
+            }
+            Throwable cause = failure.getCause();
+            if (cause != null && cause != failure) {
+                return isLikelyTransientNetworkingIssue(cause);
+            }
+            return false;
+        }
+
+        private static boolean isTransientClientError(int statusCode) {
+            return statusCode == HttpStatus.SC_REQUEST_TIMEOUT ||
+                    statusCode == 429; // Too many requests (not available through HttpStatus.XXX)
         }
 
         @Override

--- a/subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/ErrorHandlingModuleComponentRepositoryTest.groovy
+++ b/subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/ErrorHandlingModuleComponentRepositoryTest.groovy
@@ -16,6 +16,7 @@
 
 package org.gradle.api.internal.artifacts.ivyservice.ivyresolve
 
+import org.apache.http.conn.HttpHostConnectException
 import org.gradle.api.artifacts.component.ComponentArtifactIdentifier
 import org.gradle.api.artifacts.component.ComponentIdentifier
 import org.gradle.api.internal.artifacts.DefaultModuleIdentifier
@@ -35,6 +36,8 @@ import org.gradle.internal.resolve.result.BuildableArtifactSetResolveResult
 import org.gradle.internal.resolve.result.BuildableComponentArtifactsResolveResult
 import org.gradle.internal.resolve.result.BuildableModuleComponentMetaDataResolveResult
 import org.gradle.internal.resolve.result.BuildableModuleVersionListingResolveResult
+import org.gradle.internal.resource.transport.http.HttpErrorStatusCodeException
+import spock.lang.Shared
 import spock.lang.Specification
 import spock.lang.Subject
 import spock.lang.Unroll
@@ -44,7 +47,21 @@ class ErrorHandlingModuleComponentRepositoryTest extends Specification {
     private static final String REPOSITORY_ID = 'abc'
     def delegate = Mock(ModuleComponentRepositoryAccess)
     def repositoryBlacklister = Mock(RepositoryBlacklister)
-    def someException = new RuntimeException('Something went wrong')
+
+    @Shared
+    def runtimeError = new RuntimeException('Something went wrong')
+    @Shared
+    def connectTimeout = new SocketTimeoutException()
+    @Shared
+    def cannotConnect = new HttpHostConnectException(null, null)
+    @Shared
+    def missing = status(404)
+    @Shared
+    def clientTimeout = status(408)
+    @Shared
+    def tooManyRequests = status(429)
+    @Shared
+    def unauthorized = status(403)
 
     @Subject
     ErrorHandlingModuleComponentRepository.ErrorHandlingModuleComponentRepositoryAccess access
@@ -53,9 +70,18 @@ class ErrorHandlingModuleComponentRepositoryTest extends Specification {
         new ErrorHandlingModuleComponentRepository.ErrorHandlingModuleComponentRepositoryAccess(delegate, 'abc', repositoryBlacklister, maxRetries, backoff, 'abc')
     }
 
-    @Unroll("can list module versions (max retries = #retries)")
+    private static HttpErrorStatusCodeException status(int statusCode) {
+        new HttpErrorStatusCodeException("GET", "DUMMY", statusCode, "test") {
+            @Override
+            String toString() {
+                "status $statusCode"
+            }
+        }
+    }
+
+    @Unroll("can list module versions (max retries = #maxRetries, exception=#exception)")
     def "can list module versions"() {
-        access = createAccess(retries)
+        access = createAccess(maxRetries)
 
         given:
         def dependency = Mock(ModuleDependencyMetadata)
@@ -70,11 +96,11 @@ class ErrorHandlingModuleComponentRepositoryTest extends Specification {
         1 * delegate.listModuleVersions(dependency, result)
 
         when: 'exception is thrown in resolution'
-        retries * delegate.listModuleVersions(dependency, result) >> { throw someException }
+        effectiveRetries * delegate.listModuleVersions(dependency, result) >> { throw exception }
         access.listModuleVersions(dependency, result)
 
         then: 'resolution fails and repo is blacklisted'
-        1 * repositoryBlacklister.blacklistRepository(REPOSITORY_ID, someException)
+        1 * repositoryBlacklister.blacklistRepository(REPOSITORY_ID, { hasCause(it, exception) })
         1 * result.failed(_ as ModuleVersionResolveException)
 
         when: 'repo is already blacklisted'
@@ -86,12 +112,12 @@ class ErrorHandlingModuleComponentRepositoryTest extends Specification {
         0 * delegate._
 
         where:
-        retries << (1..3)
+        [maxRetries, exception, effectiveRetries] << retryCombinations()
     }
 
-    @Unroll("can resolve component meta data (max retries = #retries)")
+    @Unroll("can resolve component meta data (max retries = #maxRetries, exception=#exception)")
     def "can resolve component meta data"() {
-        access = createAccess(retries)
+        access = createAccess(maxRetries)
 
         given:
         def moduleComponentIdentifier = new DefaultModuleComponentIdentifier(DefaultModuleIdentifier.newId('a', 'b'), '1.0')
@@ -106,11 +132,11 @@ class ErrorHandlingModuleComponentRepositoryTest extends Specification {
         1 * delegate.resolveComponentMetaData(moduleComponentIdentifier, requestMetaData, result)
 
         when: 'exception is thrown in resolution'
-        retries * delegate.resolveComponentMetaData(moduleComponentIdentifier, requestMetaData, result) >> { throw someException }
+        effectiveRetries * delegate.resolveComponentMetaData(moduleComponentIdentifier, requestMetaData, result) >> { throw exception }
         access.resolveComponentMetaData(moduleComponentIdentifier, requestMetaData, result)
 
         then: 'resolution fails and repo is blacklisted'
-        1 * repositoryBlacklister.blacklistRepository(REPOSITORY_ID, someException)
+        1 * repositoryBlacklister.blacklistRepository(REPOSITORY_ID, { hasCause(it, exception) })
         1 * result.failed(_ as ModuleVersionResolveException)
 
         when: 'repo is already blacklisted'
@@ -122,12 +148,12 @@ class ErrorHandlingModuleComponentRepositoryTest extends Specification {
         0 * delegate._
 
         where:
-        retries << (1..3)
+        [maxRetries, exception, effectiveRetries] << retryCombinations()
     }
 
-    @Unroll("can resolve artifacts with type (max retries = #retries)")
+    @Unroll("can resolve artifacts with type (max retries = #maxRetries, exception=#exception)")
     def "can resolve artifacts with type"() {
-        access = createAccess(retries)
+        access = createAccess(maxRetries)
 
         given:
         def component = Mock(ComponentResolveMetadata)
@@ -144,11 +170,11 @@ class ErrorHandlingModuleComponentRepositoryTest extends Specification {
         1 * delegate.resolveArtifactsWithType(component, artifactType, result)
 
         when: 'exception is thrown in resolution'
-        retries * delegate.resolveArtifactsWithType(component, artifactType, result) >> { throw someException }
+        effectiveRetries * delegate.resolveArtifactsWithType(component, artifactType, result) >> { throw exception }
         access.resolveArtifactsWithType(component, artifactType, result)
 
         then: 'resolution fails and repo is blacklisted'
-        1 * repositoryBlacklister.blacklistRepository(REPOSITORY_ID, someException)
+        1 * repositoryBlacklister.blacklistRepository(REPOSITORY_ID, { hasCause(it, exception) })
         1 * result.failed(_ as ArtifactResolveException)
 
         when: 'repo is already blacklisted'
@@ -160,12 +186,12 @@ class ErrorHandlingModuleComponentRepositoryTest extends Specification {
         0 * delegate._
 
         where:
-        retries << (1..3)
+        [maxRetries, exception, effectiveRetries] << retryCombinations()
     }
 
-    @Unroll("can resolve artifacts (max retries = #retries)")
+    @Unroll("can resolve artifacts (max retries = #maxRetries, exception=#exception)")
     def "can resolve artifacts"() {
-        access = createAccess(retries)
+        access = createAccess(maxRetries)
 
         given:
         def component = Mock(ComponentResolveMetadata)
@@ -181,11 +207,11 @@ class ErrorHandlingModuleComponentRepositoryTest extends Specification {
         1 * delegate.resolveArtifacts(component, result)
 
         when: 'exception is thrown in resolution'
-        retries * delegate.resolveArtifacts(component, result) >> { throw someException }
+        effectiveRetries * delegate.resolveArtifacts(component, result) >> { throw exception }
         access.resolveArtifacts(component, result)
 
         then: 'resolution fails and repo is blacklisted'
-        1 * repositoryBlacklister.blacklistRepository(REPOSITORY_ID, someException)
+        1 * repositoryBlacklister.blacklistRepository(REPOSITORY_ID, { hasCause(it, exception) })
         1 * result.failed(_ as ArtifactResolveException)
 
         when: 'repo is already blacklisted'
@@ -197,12 +223,12 @@ class ErrorHandlingModuleComponentRepositoryTest extends Specification {
         0 * delegate._
 
         where:
-        retries << (1..3)
+        [maxRetries, exception, effectiveRetries] << retryCombinations()
     }
 
-    @Unroll("can resolve artifact (max retries = #retries)")
+    @Unroll("can resolve artifact (max retries = #maxRetries, exception=#exception)")
     def "can resolve artifact"() {
-        access = createAccess(retries)
+        access = createAccess(maxRetries)
 
         given:
         def artifact = Mock(ComponentArtifactMetadata)
@@ -219,11 +245,11 @@ class ErrorHandlingModuleComponentRepositoryTest extends Specification {
         1 * delegate.resolveArtifact(artifact, moduleSource, result)
 
         when: 'exception is thrown in resolution'
-        retries * delegate.resolveArtifact(artifact, moduleSource, result) >> { throw someException }
+        effectiveRetries * delegate.resolveArtifact(artifact, moduleSource, result) >> { throw exception }
         access.resolveArtifact(artifact, moduleSource, result)
 
         then: 'resolution fails and repo is blacklisted'
-        1 * repositoryBlacklister.blacklistRepository(REPOSITORY_ID, someException)
+        1 * repositoryBlacklister.blacklistRepository(REPOSITORY_ID, { hasCause(it, exception) })
         1 * result.failed(_ as ArtifactResolveException)
 
         when: 'repo is already blacklisted'
@@ -235,6 +261,43 @@ class ErrorHandlingModuleComponentRepositoryTest extends Specification {
         0 * delegate._
 
         where:
-        retries << (1..3)
+        [maxRetries, exception, effectiveRetries] << retryCombinations()
+    }
+
+    List<List<?>> retryCombinations() {
+        def retries = []
+        (1..3).each { ret ->
+            // no retries on runtime errors, missing resources, authentication errors
+            retries << [ret, runtimeError, 1]
+            retries << [ret, missing, 1]
+            retries << [ret, unauthorized, 1]
+            // retries on connect timeouts, too many requests, client timeouts
+            retries << [ret, connectTimeout, ret]
+            retries << [ret, cannotConnect, ret]
+            retries << [ret, tooManyRequests, ret]
+            retries << [ret, clientTimeout, ret]
+            // retries on server errors
+            if (ret < 3) {
+                // testing 1 and 2 is good enough coverage
+                (500..<600).each { code ->
+                    retries << [ret, status(code), ret]
+                }
+                // no retries on arbitrary codes
+                (100..<200).each { code ->
+                    retries << [ret, status(code), 1]
+                }
+            }
+        }
+        retries
+    }
+
+    static boolean hasCause(Throwable actual, Throwable cause) {
+        if (actual == cause) {
+            return true
+        }
+        if (actual.cause && actual.cause != actual) {
+            return hasCause(actual.cause, cause)
+        }
+        return false
     }
 }

--- a/subprojects/resources-http/src/main/java/org/gradle/internal/resource/transport/http/HttpErrorStatusCodeException.java
+++ b/subprojects/resources-http/src/main/java/org/gradle/internal/resource/transport/http/HttpErrorStatusCodeException.java
@@ -35,4 +35,8 @@ public class HttpErrorStatusCodeException extends RuntimeException {
     public boolean isServerError() {
         return statusCode >= 500 && statusCode < 600;
     }
+
+    public int getStatusCode() {
+        return statusCode;
+    }
 }

